File: backends/gpu/kernels/activation.ts
import { GPUTensor, Rank, Shape, Shape3D } from "../../../core/types.ts";
import { WebGPUBackend } from "../../../deps.ts";

export async function feedforward(
  backend: WebGPUBackend,
  inputs: GPUTensor<Rank>,
  outputs: GPUTensor<Rank>,
  activation: string,
) {
  const shape = to3D(inputs.shape)
  const code = shader_ff(activation, shape);
  const pipeline = await backend.register(code);
  backend.execute(
    pipeline,
    [Math.ceil(shape[0] / 64), shape[1], shape[2]], 
    [inputs.data, outputs.data]
  );
}

const shader_ff = (activation: string, shape: Shape3D) => 
`struct Matrix {
  values: array<f32>
};

@group(0) @binding(0)
var<storage, read> inputs: Matrix;
@group(0) @binding(1)
var<storage, write> outputs: Matrix;

@compute @workgroup_size(64, 1, 1)
fn main(@builtin(global_invocation_id) global_id: vec3<u32>) {
  if (
    global_id.x < ${shape[0]}u && 
    global_id.y < ${shape[1]}u && 
    global_id.z < ${shape[2]}u
  ) {
    var idx = global_id.x + global_id.y * ${shape[0]}u + global_id.z * ${shape[1]}u;
    ${activation};
  }
}`;

export async function backpropagate(
  backend: WebGPUBackend,
  inputs: GPUTensor<Rank>,
  dError: GPUTensor<Rank>,
  dInputs: GPUTensor<Rank>,
  prime: string,
) {
  const shape = to3D(inputs.shape)
  const code = shader_bp(prime, shape);
  const pipeline = await backend.register(code);
  backend.execute(
    pipeline,
    [Math.ceil(shape[0] / 64), shape[1], shape[2]], 
    [inputs.data, dError.data, dInputs.data],
  );
}

const shader_bp = (prime: string, shape: Shape3D) => 
`struct Matrix {
  values: array<f32>
};

@group(0) @binding(0)
var<storage, read> inputs: Matrix;
@group(0) @binding(1)
var<storage, read> dError: Matrix;
@group(0) @binding(2)
var<storage, write> dInputs: Matrix;

@compute @workgroup_size(64, 1, 1)
fn main(@builtin(global_invocation_id) global_id: vec3<u32>) {
  if (
    global_id.x < ${shape[0]}u && 
    global_id.y < ${shape[1]}u && 
    global_id.z < ${shape[2]}u
  ) {
    var idx = global_id.x + global_id.y * ${shape[0]}u + global_id.z * ${shape[1]}u;
    ${prime};
  }
}`;

export const sigmoid =
  `outputs.values[idx] = 1.0 / (1.0 + exp(-inputs.values[idx]));`;

export const sigmoid_prime =
  `var activation = inputs.values[idx] * (1.0 - inputs.values[idx]);
dInputs.values[idx] = activation * dError.values[idx];`;

function to3D(shape: Shape[Rank]): Shape3D {
  let res = shape as Shape3D
  switch (shape.length) {
    case 2:
      res = [shape[0] * shape[1], 1, 1];
      break
    case 4:
      res = [shape[0] * shape[1], shape[2], shape[3]];
  }
  return res;
}

File: crates/core-gpu/src/gpu/layers/activation.rs
use ndarray::{Dimension, IxDyn};

use crate::{ActivationLayer, GPUActivation, WGPUBackend, WGPUBuffer, WGPUKernel};

pub struct ActivationGPULayer {
    // data
    pub memoize_output: bool,
    pub outputs: WGPUBuffer,

    // gradients
    pub d_inputs: WGPUBuffer,

    // kernels
    pub forward_kernel: WGPUKernel,
    pub backward_kernel: WGPUKernel,
}

impl ActivationGPULayer {
    pub fn new(backend: &mut WGPUBackend, config: ActivationLayer, size: &mut IxDyn) -> Self {
        let activation = GPUActivation::from(config.activation);
        let forward_kernel = kernel_forward(backend, size.size(), activation.activate);
        let backward_kernel = kernel_backward(backend, size.size(), activation.prime);

        Self {
            memoize_output: GPUActivation::memoize_output(&activation.activation),
            outputs: WGPUBuffer::new(backend, size.clone()),
            d_inputs: WGPUBuffer::new(backend, size.clone()),
            forward_kernel,
            backward_kernel,
        }
    }

    pub fn reset(&mut self, backend: &mut WGPUBackend, batches: usize) {
        let output_size = self.outputs.shape.as_array_view()[1];
        self.outputs = WGPUBuffer::new(backend, IxDyn(&[batches, output_size]))
    }

    pub fn forward_propagate(&self, backend: &mut WGPUBackend, inputs: &WGPUBuffer) {
        backend.execute(&self.forward_kernel, vec![inputs, &self.outputs]);
    }

    pub fn backward_propagate(
        &self,
        backend: &mut WGPUBackend,
        inputs: &WGPUBuffer,
        d_outputs: &WGPUBuffer,
    ) {
        if self.memoize_output {
            backend.execute(
                &self.backward_kernel,
                vec![&self.outputs, d_outputs, &self.d_inputs],
            );
        } else {
            backend.execute(
                &self.backward_kernel,
                vec![inputs, d_outputs, &self.d_inputs],
            );
        };
    }
}

fn kernel_forward(backend: &mut WGPUBackend, size: usize, activation: String) -> WGPUKernel {
    let source = format!(
        "struct Matrix {{
            values: array<f32>
        }};
          
        @group(0) @binding(0)
        var<storage, read> inputs: Matrix;
        @group(0) @binding(1)
        var<storage, read_write> outputs: Matrix;
          
        @compute @workgroup_size(64, 1, 1)
        fn main(@builtin(global_invocation_id) global_id: vec3<u32>) {{
            if (global_id.x < {size}u) {{
                var x = inputs.values[global_id.x];
                outputs.values[global_id.x] = {activation};
            }}
        }}"
    );
    backend.register(source, ((size as f64 / 64.0).ceil() as u32, 1, 1))
}

fn kernel_backward(backend: &mut WGPUBackend, size: usize, activation: String) -> WGPUKernel {
    let source = format!(
        "struct Matrix {{
            values: array<f32>
        }};
          
        @group(0) @binding(0)
        var<storage, read> inputs: Matrix;
        @group(0) @binding(1)
        var<storage, read> d_outputs: Matrix;
        @group(0) @binding(2)
        var<storage, read_write> d_inputs: Matrix;
          
        @compute @workgroup_size(64, 1, 1)
        fn main(@builtin(global_invocation_id) global_id: vec3<u32>) {{
            if (global_id.x < {size}u) {{
                var d_output = d_outputs.values[global_id.x];
                var x = inputs.values[global_id.x];
                d_inputs.values[global_id.x] = {activation} * d_output;
            }}
        }}"
    );
    backend.register(source, ((size as f64 / 64.0).ceil() as u32, 1, 1))
}
